helix-term/src/ui/picker.rs

```diff
@@ -946,21 +1071,98 @@ impl<T: Item> Drop for Picker<T> {
 
 type PickerCallback<T> = Box<dyn Fn(&mut Context, &T, Action)>;
 
+fn parse_query(
+    column_names: &[&'static str],
+    primary_column: usize,
+    input: &str,
+) -> HashMap<&'static str, String> {
+    let mut fields: HashMap<&'static str, String> = HashMap::new();
+    let primary_field = column_names[primary_column];
+    let mut escaped = false;
+    let mut quoted = false;
+    let mut in_field = false;
+    let mut field = None;
+    let mut text = String::new();
+
+    macro_rules! finish_field {
+        () => {
+            let key = field.take().unwrap_or(primary_field);
+
+            if let Some(pattern) = fields.get_mut(key) {
+                pattern.push(' ');
+                pattern.push_str(&text);
+                text.clear();
+            } else {
+                fields.insert(key, std::mem::take(&mut text));
+            }
+        };
+    }
+
+    for ch in input.chars() {
+        match ch {
+            // Backslash escaping
+            '\\' => escaped = !escaped,
+            _ if escaped => {
+                // Keep backslash escapes for special nucleo syntax.
+                if matches!(ch, '^' | '$' | '!' | '\'') {
+                    text.push('\\');
+                }
+                text.push(ch);
+                escaped = false;
+            }
+            // Double quoting
+            '"' => quoted = !quoted,
+            ' ' if quoted => {
+                text.push('\\');
+                text.push(' ');
+            }
+            '%' | ':' if quoted => text.push(ch),
+            // Space either completes the current word if no field is specified
+            // or field if one is specified.
+            '%' | ' ' if !text.is_empty() => {
+                finish_field!();
+                in_field = ch == '%';
+            }
+            '%' => in_field = true,
+            ' ' => (),
+            ':' if in_field => {
+                // Go over all columns and their indices, find all that starts with field key,
+                // select a column that fits key the most.
+                field = column_names
+                    .iter()
+                    .filter(|col| col.starts_with(&text))
```

Comment:
@pascalkuthe what do you think about using fuzzy matching for this instead? It would be more forgiving, so you could use `%pth:syntax.rs` for `path` for example. We could even highlight the parts of the column name that matched, like we do for the parts of the query that match items

Reply:
We discussed this and fuzzy matching seems like overkill - the prefix is more straightforward. Highlighting the matching parts could also be done for prefix strings as well and would be simpler. (Could be a follow-up to #9647)